controllers/machineset_controller.go

```diff
@@ -126,21 +127,41 @@ func (r *ReconcileMachineSet) ProcessMachineSet(ctx context.Context, machineSet
 		// Update labels in machine
 		err = r.updateLabelsInMachine(ctx, machine, expectedLabels)
 		if err != nil {
+			metrics.RegisterFailedLabelUpdate(&metrics.MnmoLabels{
```

Comment:
I think we need some more information in these metrics to distinguish past issues from ongoing issues.  Ideally we would be able to form a  promql query/alert such as "the latests machinestate is successfully applied within N minutes".

For example, say a customer starts with machinestate A and updates it to machinestate B to add some labels, but machinestate B is broken for some reason so these error metrics fire for a bit. The customer then fixes it in machinestate C so the error metrics are no longer firing.

Looking at the MachineSet, i see a [Generation](https://github.com/kubernetes/apimachinery/blob/7300632adf3471b534f287645ac2acb6ecfc94f2/pkg/apis/meta/v1/types.go#L179) in ObjectMetadata which sounds promising as a version sequence. Theres also a status which has an [ObservedGeneration](https://github.com/openshift/cluster-api/blob/bd5518e2878206372eb4cd24e3f03a2bb82b6d1d/pkg/apis/machine/v1beta1/machineset_types.go#L146) field. Note sure what the relationship between these two are exactly but naiively it sounds like ObservedGeneration might be the currently applied generation while Generation is the currently generation we are reconciling.

Reply:
We do need to be mindful of cardinality of the metrics. Adding some kind of version as a label will cause the cardinality to go through the roof imho, as there will be a new timeseries created everytime a customer changes a machineset.

What you are describing can also just as easily be detected by looking at a counter that counts successful and unsuccessful reconciliations.

Reply:
> as there will be a new timeseries created everytime a customer changes a machineset.
Yes, that is what i want. Metrics associated with customer use of this feature, both successful and not. Disk space is cheap.